### ISSUE TITLE
Add high DPI support for Windows 10

### ISSUE THREAD

**Issue by daniel-white** [Microsoft/winfile#19]

**Comment by jhudsoncedaron** [comment]
So I discovered that yeah it doesn't know about DPI; however a naive implementation would break me as my DPI is 192! Since this component doesn't invoke any browser controls I find the missing high DPI support a non-issue.

**Comment by wojtekmaj** [comment]
I think it's part of the fun that we can count pixels in this :D

**Comment by daniel-white** [comment]
Yea @jhudsoncedaron i'm at the same level as you. 

i hate the pixelated text. it looks like we would just have to provide new resources:

![image](https://user-images.githubusercontent.com/735269/38528259-cf331a0e-3c2c-11e8-9a81-81ae33e5b1e7.png)

**Comment by wojtekmaj** [comment]
On a side note: In Windows 10, if you select System (enhanced) DPI handling in Compatibility settings, it works beautifully.

**Comment by sean-e** [comment]
consider adding an app manifest with this:
```
<asmv3:application>
	<asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2017/WindowsSettings">
		<gdiScaling>true</gdiScaling>
	</asmv3:windowsSettings>
</asmv3:application>
```

**Comment by craigwi** [comment]
@daniel-white, do the suggestions from @wojtekmaj  or @sean-e work for you?

**Comment by My1** [comment]
I have to disagree with the title.

it just should be "add high DPI support". Win7 and 8 are also capable of DPI scaling so why try to artificially limit it?

**Comment by daniel-white** [comment]
@craigwi it would be nice if it was built in. @sean-e's suggestion would work, but its not per monitor aware like how windows 10 supports. I personally used `SetProcessDpiAwareness(PROCESS_PER_MONITOR_DPI_AWARE)` in `WinMain`. No problem there.

@My1 windows 10's is superior. it has per monitor detection, so set the bar high and i think 7/8 will come with less work or free.

**Comment by My1** [comment]
@daniel-white win8 at least (I didnt try 7) can also use different scalings per monitor, I dont remember whether it auto detects, but I honestly dont care, the important thing is that this thing can at least properly scale on all OSes that can do scaling.

**Comment by craigwi** [comment]
@daniel-white the change seems simple (adding the API call).  I don't have a high DPI environment.  It sounds like you have verified this solution works.  Have you verified it doesn't cause problems in "normal" environments?

**Comment by daniel-white** [comment]
@craigwi i only tested locally on my machine. i only have hi-dpi machines. im guessing its extra `WM_` that need handled, backwards compatibility checks and hi-res resources.

**Comment by NazmusLabs** [comment]
@creigwi good news is that you don't need a high DPI device to test in high a high DPI environment. You can easily change the display scaling from the Windows 10's settings app to achieve the same effect. One can then do any high DPI related testing and then switch back to 100% scaling. I, for example, being visually impared, have set my display scaling to 150% on my 27' 1080p monitor, even though most would keep scaling at 100% on a screen of this size. As a result, the OS and the apps act as though I am running in an high DPI environment.

![high dpi settings on a non high dpi display for visual aid](https://user-images.githubusercontent.com/5572377/39006121-269afc6c-43d0-11e8-92e4-1390fafad78f.png)

P.S. for legacy apps (which WinFile is) it's necessary to sign out and sign back in again to Windows after changing display scaling so as to not have the app look blurry. UWP apps are unaffected.

**Comment by craigwi** [comment]
Thanks @NazmusLabs.  I guess the next step is to test the API call that @daniel-white  suggests.

**Comment by NazmusLabs** [comment]
@daniel-white I have been trying to include `SetProcessDpiAwareness(PROCESS_PER_MONITOR_DPI_AWARE)` inside the `winmain` function of winfile.c, however, visual studio is not happy at all at compilation. I have tried different ways to including that API call with no avail. What did you do that I am forgetting that is preventing me from compiling the code?

**Comment by peterooch** [comment]
@NazmusLabs 
I think you need to put that in InitFileManager function,  which does all the initialization.

**Comment by My1** [comment]
@craigwi can you send over the exe? I can help testing.

**Comment by craigwi** [comment]
Attached.
[Debug#19.zip](https://github.com/Microsoft/winfile/files/1944234/Debug.19.zip)

**Comment by craigwi** [comment]
https://www.microsoft.com/en-us/download/details.aspx?id=48145

**Comment by My1** [comment]
I have that already., iirc it's release vs debug building.

**Comment by craigwi** [comment]
Release build attached.
[Release#19.zip](https://github.com/Microsoft/winfile/files/1944297/Release.19.zip)

**Comment by My1** [comment]
I think this can count as an almost pass.  (toolbars need some scaling)

but the text looks great

and most importantly, it is running on versions <w10 (8.1 in my case)

![screenshot 142](https://user-images.githubusercontent.com/6696524/39209909-7d9cc7f8-4807-11e8-9035-15eba306a4fd.png)

**Comment by NazmusLabs** [comment]
![high dpi set to application](https://user-images.githubusercontent.com/5572377/39210957-3bf259e6-47d8-11e8-8996-06a5a67c72fe.png)

@My1 I can confirm. I'm having the same issue in Windows 10. So it's not a Windows 7 thing.

@craigwi so, apparently it didn't work. The result is the same as if we set DPI to "Application" instead of the default "system" in the Windows compatibility tab. I'll need to look through the docs.

**Comment by NazmusLabs** [comment]
@craigwi I figured out what happened. The manifest tells Windows that the application is DPI aware, so Windows will stop scaling the application, as it does for DPI _unware_ apps. But the thing is, this is just a metadata. Winfile isn't _actually_ DIP aware. The manifest just _tells_ Windows that it is so that Windows won't interfere with the application's scaling code.


So, in summery, it appears that this manifest info is for application that has its own high DPI scaling code. Winfile doesn't have its own scaling code; so, this doesn't apply here.

**Comment by My1** [comment]
@NazmusLabs but how can it NOT have scaling is all the text elements DO scale beautifully?

**Comment by ZanderBrown** [comment]
@My1 toolbar should be relatively easy but will require 1x1px -> 2x2px treatment on the icons but I believe the drivebar is a little more 'custom' so could be a can of worms

**Comment by NazmusLabs** [comment]
And, folks, I think I have found what we are looking for:

We need to target the _newer_ system enhanced scaling model, and I've found a sourceforge Q/A page about how to do this. It seems simple, as expected, since we are asking Windows to do all the hard work. We just need to have the right manifest.

This is like Visual Styles all over again. We just need to target the later version than what we are originally doing.

Here's the page: https://stackoverflow.com/questions/46428510/enhanced-system-dpi-scaling-with-vs2017

One thing I want to note is that it seems to imply that this will work in VS2017 (which make sense, as this feature JUST came out a few months ago, i.e. late 2017). I am doubtful it will make it to VS2015. But correct me if I'm wrong.

@ZanderBrown we might be able to solve this with the System Enhanced. It looks orders of magnitude better (but not as awesome as true scaling, as you can see with the Frankenstein's parts where scaling works). This will involve almost no work on our part. You can try it out now by setting the DPI settings on Winfile to System Enhanced (right-click, properties, compatibility, DPI settings).

**Comment by My1** [comment]
@NazmusLabs but pls try not to force sys-enhanced scaling, as this will completely kick w7/8.1 users. also if the scaling gets crazy (300%+ maybe already, which may be needed on some screens or just if you cant see that well and want stuff bigger) even sys-enhanced will probably hit limits.

**Comment by NazmusLabs** [comment]
Here's a screenshot of how the system enhanced looks. I'm currently running at 150% scaling. Notice how much of an improvement it makes. You will see System Enhanced vs the "Frankenstein" version. As you may note, that the part of the "Frankenstein" one that DOES scale (Help Menu in this example) looks better than System Enhanced. But System Enhanced does offer a better _overall_ look.

![winfile frankenstein vs system enhanced 2](https://user-images.githubusercontent.com/5572377/39212402-a7baf62a-47dc-11e8-8c1e-40cedcf124be.png)

![winfile frankenstein vs system enhanced 1](https://user-images.githubusercontent.com/5572377/39212401-a7a5fd60-47dc-11e8-8608-56a6544e710d.png)

**Comment by My1** [comment]
certainly, although if we can fix both icon bars, by adapting them to the new UI elements) that would also make sys-enhanced obsolete

**Comment by NazmusLabs** [comment]
@my1 Don't worry, using system enhanced will have no effect on Windows 7. It'll just improve Windows 10 (if we do things correctly, as we did with visual styles). Using it on Windows 7 won't cause a regression. You just will see it act the way it does right now. Scaled but blurry. But you shouldn't see anything get WORSE.

Having system enhanced is better than having nothing at all. And if/when we eventually are able to migrate to the latest Win32 controls, Windows 7 users can definitely benefit. So until we get pure high DPI support, might as well make use of a Windows 10 feature, since it doesn't hurt anyone.

**Comment by ZanderBrown** [comment]
@My the dropdown is 'new but the 'new' toolbar it's in is only as big as it's tallest button which is based on the icon sizes which are defined as bitmaps [here](https://raw.githubusercontent.com/Microsoft/winfile/master/src/toolbar.bmp) where they are 15px tall

**Comment by My1** [comment]
sure it wont go worse, but it wont get better.

reading the second paragraph, it sounds liek a way, doing it as an intermediate sounds like a not too crazy way to go.

but is migrating to new controls that hard for the toolbars?

**Comment by ZanderBrown** [comment]
for the maintoolbar just having bigger icons should do the trick, but then we need to workout which size to use

**Comment by NazmusLabs** [comment]
@My1 yeah, @ZanderBrown is correct. See, recall the time when I made a post about the "dark" theme Winfile in Windows 3.1? Every part of the app looked and worked beautifully, EXCEPT the drive bar. This is exactly what happens when we stray away from native controls. The more native we stay, the easier it is to make the app forward _and_ backword compatible.


A great example: all Win32 apps using NATIVE titlebars will get "sets" feature, according to the latest insider build. So, I'm very excited (and have tested it working beautifully) about Winfile getting tabs support for "free". All it had to do is stick native, and it's getting a new UI feature. Other apps, like Spotify, or Word 2016, is using non native title bars and won't get tabs. 😢

**Comment by NazmusLabs** [comment]
@My1 yeah, @ZanderBrown is correct. See, recall the time when I made a post about the "dark" theme Winfile in Windows 3.1? Every part of the app looked and worked beautifully, EXCEPT the drive bar. This is exactly what happens when we stray away from native controls. The more native we stay, the easier it is to make the app forward _and_ backword compatible.


A great example: all Win32 apps using NATIVE titlebars will get "sets" feature, according to the latest insider build. So, I'm very excited (and have tested it working beautifully) about Winfile getting tabs support for "free". All it had to do is stick native, and it's getting a new UI feature. Other apps, like Spotify, or Word 2016, is using non native title bars and won't get tabs. 😢

**Comment by NazmusLabs** [comment]
@ZanderBrown just tried making the icons for the toolbars 4 times larger, pixel for pixel. They just get cut off and the toolbar size doesn't change. I didn't change any code, just increased the bimap resolution by 4x.

**Comment by ZanderBrown** [comment]
@NazmusLabs custom isn't necessarily bad as long as your controls does the right thing and instead of hardcoding colours uses the likes of `GetSysColor`, of course the irony with custom titlebars is that MS actively encouraged it with Vista

**Comment by My1** [comment]
@NazmusLabs your big comment popped in twice.

@ZanderBrown well bigger icons sure, regarding size I think of 2 ways:

1) vector
2) create as vector, export at multiple common ones (1,5x 2x 3x and maybe some more)

@NazmusLabs 
but why isnt the drivebar native? it's basically just a toolbar with letters next to the buttons and highlighting of these buttons, I think office 2003 or so already also had that and I doubt it wouldnt do scaling.

but that word 2016 isnt native is kinda crazy and this also shows why I hate it when applications start with non-native design stuff, just stay native, also since this allows modders to change styles with ease (like me having a win7 style on 8.1)

@ZanderBrown why did MS encourage non-native titlebars with vista?

**Comment by NazmusLabs** [comment]
@ZanderBrown True. I guess I should rephrase my point. Use native guidelines. :P and now that you mentioned colors, unfortunately, that's exactly the kind of trap Google fell into. They literally HARD CODED the text selection color in chrome to match the text selection color that Windows 7 uses. No one noticed until Windows 10 Creators update, which changed the text scaling color to the new Windows 10 blue rather than Windows 7 cyan. Now, firefox, and Edge has the correct highlighting and Chrome has become a Frankenstein. On chrome, the address bar text selection color is native and matches Windows 10's color. But the web page selection color was hard coded and is still using the Windows 7 color. This has been driving me nuts for over a year since no one at google seems to catch this glaring oversight. How do they miss this stuff for over a year is beyond me.

**Comment by ZanderBrown** [comment]
@NazmusLabs my bad you also need to send [TB_SETBITMAPSIZE](https://msdn.microsoft.com/en-us/library/windows/desktop/bb787398(v=vs.85).aspx) in CreateFMToolbar

@My1 yes I understand the point about vectors but this is Win32, so 24bit bitmap is what we get

**Comment by My1** [comment]
@ZanderBrown well then as said, option two: create as vector and export in different sizes as needed before building.

but it's kinda hilarious that a system that is supposed to do scaling and stuff doesnt do vectors natively for icons

**Comment by NazmusLabs** [comment]
@My1 The Win32 controls doesn't support vectors. Win32 apps uses the GDI controls that dates back to early 90s with Windows 3.0 if not older. At the time, all UI elements used pixel based bitmap graphics or raster text UI. And even today, since the Win32 controls are based on the bitmap based GDI controls, it cannot move away from that. This is why Microsoft has moved to XAML with UWP. UWP apps uses DirectX and vector based rendering for its user interfaces. They cannot do that with Win32 controls without breaking the older apps.

(personal note, they could certainly make the XAML based UI accessible to legacy Win32 apps so existing apps can use those instead of the old GDI. I have always been against Microsoft's decision to only keep the XAML, vector based, UI exclusive to UWP apps. But that's another topic of discussion not relevant here.

**Comment by ZanderBrown** [comment]
@NazmusLabs something like `SendMessage(hwndToolbar, TB_SETBITMAPSIZE, 0, MAKELPARAM(64, 60) );` around line 1111 of tbar.c

@My1 considering we are working with 80s tech it isn't supposed to do vectors, that's very much an after thought

**Comment by My1** [comment]
@NazmusLabs I would have thought it would be possible without breaking everything.
like if we get a vector icon, we use it, and if it's bitmap we use that, or maybe bring it as part of .net that thing didnt exist THAT long.

**Comment by NazmusLabs** [comment]
@ZanderBrown yeah, but @My1 has a clear point that I agree with. Windows DOES HAVE vector based counterparts for almost ALL win32 GDI controls. Microsoft is CHOOSING to make them inaccessible for Win32. @My1 asked a fair question. Why not offer vector UI. Microsoft would have had an excuse if Windows didn't have any vector based controls and could say it'll be some time before they are created. But they exist. Perhaps, it's technically extremely difficult, but I find that hard to believe.

**Comment by ZanderBrown** [comment]
@NazmusLabs @My1 good points but we could be here forever discussing annoyances/stupidities of Windows API design but for now we are trying to get something done with what we've got to work with

**Comment by My1** [comment]
but even if they dont offer vectors, we could at least create in vector and then export in  pixel to solve the hassle of re-creating them for each res.

where are the icons, so I can try to whip something together? didnt find them in the source.

**Comment by ZanderBrown** [comment]
@My1 nope that's how classic Win32 worked

**Comment by My1** [comment]
well lol, doesnt make it easier but no problem.

here some WIP on the last 2 icons, the borders are obviously only for me working on this. on export they are gone.

does w32 at least eat an alpha channel properly, or do I need to go full old school and assign some crazy color as transparent?

![screenshot 143](https://user-images.githubusercontent.com/6696524/39215687-15d44558-4819-11e8-9976-61ea66e453ad.png)
while I work with a base height of 64px, it's mostly meaningless to vector anyway.

**Comment by ZanderBrown** [comment]
Win32 uses 24bit so Alpha channel will cause all kinds of chaos :smiley: , best bet use the background of the original

Considering we are generally followin gthe Win3x style colour in toolbar icons doesn't seem appropriate but the 9x style 'hot' toolbar could be a fun option

**Comment by My1** [comment]
maybe the toolbar needs to be told that the bitmap is this and that format.

**Comment by ZanderBrown** [comment]
For higher res it's looking okay

**Comment by NazmusLabs** [comment]
I love how it started to the monochrome clean icons in the early 90s, then a bit of color in the late 90s, to fisher price icons in the 2000s to glossy ones and now, we aback to flat, clean monochrome icons with Windows 10. Winfile was so late to the party, it's fashion became outdated to ancient and back to being trendy again. 😅

**Comment by clzls** [comment]
Is anybody still working on this topic now? @My1 or anybody else?

**Comment by kowalski7cc** [comment]
On Windows 1904 it looks really blurry
